File: javascript/memory_example.js
/**
 * Memory round-trip example.
 *
 * Stores a short fact, retrieves it by semantic query, shows the top hit.
 *
 * Run:
 *     npm install @pathcourse/sdk
 *     PCH_API_KEY=pch_prod_b_... node javascript/memory_example.js
 */

import { randomUUID } from 'node:crypto';
import { PathCourseClient } from '@pathcourse/sdk';

async function main() {
  const apiKey = process.env.PCH_API_KEY;
  if (!apiKey) { console.error('PCH_API_KEY env var required.'); process.exit(1); }

  const client = new PathCourseClient({ apiKey });

  const marker  = randomUUID().slice(0, 8);
  const content = `Memory test [${marker}]: the capital of Japan is Tokyo.`;

  console.log('store  :', JSON.stringify(content));
  const storeR = await client.memory.store({
    content,
    memoryType: 'semantic',   // episodic | semantic | procedural | working | shared | multimodal
    namespace:  'private',
    importance: 0.8,
  });
  console.log('  memory_id        :', storeR.memory_id);
  console.log('  balance_remaining:', storeR.balance_remaining);

  const retr = await client.memory.retrieve({
    query:     'What is the capital of Japan?',
    namespace: 'private',
    topK:      3,
  });
  const hits = retr.results || [];
  console.log(`\nretrieve: ${hits.length} hit(s)`);
  for (const h of hits.slice(0, 3)) {
    console.log(
      `  - score=${h.similarity_score}  id=${(h.memory_id || '').slice(0, 8)}...  ` +
      `content_preview=${JSON.stringify(h.content_preview)}`
    );
  }

  // Memory retrieve returns content_preview (first 200 chars) — not full
  // stored text. Full content is only used to generate the embedding.
  console.log('\ncost this request:', retr.cost_usdc, 'USDC');
}

main().catch((err) => { console.error(err); process.exit(1); });

File: python/memory_example.py
"""
Memory round-trip example.

Stores a short fact, retrieves it by semantic query, and verifies the top
hit contains the content we stored. Demonstrates the embedding-based
persistent memory layer every PCH agent gets by default.

Run:
    pip install pathcourse-sdk
    export PCH_API_KEY=pch_prod_b_...
    python python/memory_example.py
"""

import os
import sys
import uuid

from pathcourse import PathCourseClient


def main() -> int:
    key = os.environ.get("PCH_API_KEY")
    if not key:
        print("PCH_API_KEY env var required.", file=sys.stderr)
        return 1

    client = PathCourseClient(api_key=key)

    # Unique marker so repeated runs don't hit the content_hash dedup path.
    marker = uuid.uuid4().hex[:8]
    content = f"Memory test [{marker}]: the capital of Japan is Tokyo."

    print(f"store  : {content!r}")
    store_r = client.memory.store(
        content=content,
        memory_type="semantic",   # episodic | semantic | procedural | working | shared | multimodal
        namespace="private",
        importance=0.8,
    )
    print(f"  memory_id         : {store_r['memory_id']}")
    print(f"  balance_remaining : {store_r.get('balance_remaining')}")

    retr = client.memory.retrieve(
        query="What is the capital of Japan?",
        namespace="private",
        top_k=3,
    )
    hits = retr.get("results", [])
    print(f"\nretrieve: {len(hits)} hit(s)")
    for h in hits[:3]:
        print(f"  - score={h.get('similarity_score')}  id={h.get('memory_id', '')[:8]}...  "
              f"content_preview={h.get('content_preview')!r}")

    # Memory lookup returns content_preview (first 200 chars) — not the full
    # stored text. Full content is only used to generate the embedding.
    print(f"\ncost this request : {retr.get('cost_usdc')} USDC")
    return 0


if __name__ == "__main__":
    sys.exit(main())
